[DO NOT MERGE] Epoch manager WIP

martinvol opened the pull request:
Placeholder PR to have the CI

soloseng committed:
first compiling draft
no working test.

Co-authored-by: Martín Volpe <martinvol@users.noreply.github.com>

pahor167 committed:
Compiler fixes + finishNextEpochProcess

pahor167 committed:
allocateValidatorsRewards compilable

pahor167 committed:
First few tests

pahor167 committed:
test fix

soloseng committed:
uisng registry instead

soloseng committed:
updated registry ID var names

soloseng committed:
typo

soloseng committed:
++ mock contracts in 0.8

soloseng committed:
++ passing test using mock

pahor167 committed:
Removal of using precompiles

pahor167 committed:
ScoreManager refactor

pahor167 committed:
Decouple epoch manager initializer from registry

pahor167 committed:
isReadyToStartEpoch fix

pahor167 committed:
startNextEpochProcess fixes

pahor167 committed:
rename score manager to score reader

pahor167 committed:
Rename transfer to release on CeloDistribution schedule

pahor167 committed:
Celo distribution schedule renamed to CeloUnreleasedTreasure

pahor167 committed:
conditions to getFirstBlockAtEpoch and getLastBlockAtEpoch

pahor167 committed:
ScoreManager update

pahor167 committed:
getFirstBlockOfEpoch in EpochManager initializer

pahor167 committed:
extra checks

pahor167 committed:
introduce new getValidatorsGroup

pahor167 committed:
systemAlreadyInitialized update

pahor167 committed:
Removal of IEpochManagerInterface

martinvol committed:
Made CI run on feature branch

pahor167 committed:
Merge branch 'release/core-contracts/12' into feat/epoch-manager

pahor167 committed:
merge fix

martinvol committed:
Added target for the CI

martinvol committed:
Merge branch 'feat/epoch-manager' of github.com:celo-org/celo-monorepo into feat/epoch-manager

martinvol committed:
Merge branch 'feat/epoch-manager' into feat/l2-epoch-system

martinvol committed:
Fixed quote celo-monorepo.yml

m-chrzan committed:
Allow Validator registration in L2 without BLS key (#11181)

* Allow validator registration in L2 without BLS key

* Reallow ECDSA key change in L2

---------

Co-authored-by: Martín Volpe <volpe@clabs.co>

martinvol committed:
Fix CI

martinvol committed:
Fixed interfaces and imports

pahor167 committed:
Build fix

pahor167 committed:
lint

pahor167 committed:
prettify

pahor167 committed:
prettify 2

martinvol committed:
Attempt to fix truffle build and migrations l2 epoch (#11190)

martinvol committed:
Added Blocking to LockedGold, Election (#11186)

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManagerInitializer.sol:
```diff
@@ -0,0 +1,62 @@
+    return _getFirstBlockOfEpoch(currentEpoch);
+  }
+
+  function _getFirstBlockOfEpoch(uint256 currentEpoch) internal view returns (uint256) {
```
this is quite suboptimal.

Either do a binary search to check if an block is what epoch, or check `pos.go` to implement this functions

There's a solidity version of epochSize and with that you can compute epoch for block X pretty easy

From pos.go
```
// GetEpochNumber retrieves the epoch number given the block number.
// Epoch 0 is a special block that only contains the genesis block (block 0), epoch 1
// starts at block 1
func GetEpochNumber(number uint64, epochSize uint64) uint64 {
	if IsLastBlockOfEpoch(number, epochSize) {
		return number / epochSize
	} else {
		return number/epochSize + 1
	}
}

// MustGetEpochFirstBlockGivenBlockNumber is a variant of GetEpochFirstBlockGivenBlockNumber
// that panics if called for epoch 0 (genesis)
func MustGetEpochFirstBlockGivenBlockNumber(blockNumber uint64, epochSize uint64) uint64 {
	firstBlock, err := GetEpochFirstBlockGivenBlockNumber(blockNumber, epochSize)
	if err != nil {
		panic(err)
	}
	return firstBlock
}

// GetEpochFirstBlockGivenBlockNumber retrieves first block of a given block's epoch
// Fails when try to obtain first block of epoch 0 (genesis)
func GetEpochFirstBlockGivenBlockNumber(blockNumber uint64, epochSize uint64) (uint64, error) {
	epochNumber := GetEpochNumber(blockNumber, epochSize)
	return GetEpochFirstBlockNumber(epochNumber, epochSize)
}

// GetEpochFirstBlockNumber retrieves first block of epoch.
func GetEpochFirstBlockNumber(epochNumber uint64, epochSize uint64) (uint64, error) {
	// Epoch 0 is just the genesis block, it doesn't have a first block (only last)
	if epochNumber == 0 {
		return 0, errors.New("No first block for epoch 0")
	}

	return ((epochNumber - 1) * epochSize) + 1, nil
}
```

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManagerInitializer.sol:
```diff
@@ -0,0 +1,62 @@
+  /**
+   * @notice initializes the epochManager contract during L2 transition.
+   */
+  function initEpochManager() external onlyOwner {
```
should be split into 2 functions:

1. `captureEpochAndValidators()` which is permissionless and can be called while `isOnL2` is false. This function captures: validators and epoch number
2. `initEpochManager()` it's permissionless and can only be called when `isOnL2` is true. This function uses captured data and calls `EpochManager.initilizae()`

Then

Function `initEpochManager()` can only be called once, since EpochManager won't allow 2 calls to `initialize()`

soloseng replied:
Created an [issue](https://github.com/celo-org/celo-blockchain-planning/issues/590) tracking changes to the epochManagerInitializer

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManagerInitializer.sol:
```diff
@@ -0,0 +1,62 @@
+      address validatorAddress = validatorSignerAddressFromCurrentSet(i);
+      electedValidatorAddresses[i] = validatorAddress;
+    }
+    getEpochManager().initializeSystem(
```
add `startOfEpochTimestamp`

mcortesi replied:
to compute timestamp, do:

`block.timestamp - 5 * (block.number - firstBlockOfEpoch)`

mcortesi reviewed (commented):
review epochmanager initializer

pahor167 committed:
add EpochManager to abis

pahor167 committed:
Merge branch 'feat/l2-epoch-system' of https://github.com/celo-org/celo-monorepo into feat/l2-epoch-system

pahor167 committed:
force deploy abis

pahor167 committed:
build fix

martinvol committed:
Move Validators.sol to 0.8 (#11192)

* WIP

* Validator test WIP, forge doesn't compile yet

* buildable

* most of the foundry tests working

* Validator related tests fixed

* truffle build working

* cache bump

* lint + prettify+ migrations

* Enable optimization for Solidity 0.8

* prettify

* Ci bump

* CI bump 2

* Validators to 0.8 config

* CI bump

* foundry fix

* Truffle migrations are partly fixed

* Added import for ValidatorsMock08 in EpochManager.t.sol, I think it was removed by mistake

* Added yarn.lock

* Changes to mock with full implementation

* Attempt to fix linking libraries not working with deployCodeTo https://github.com/foundry-rs/foundry/issues/4049

* truffle migrations fixed

* CI bump

* lint

* forge test fixes

* artifacts test fix

* lint

* update of foundry version

* add ProxyFactory import to tests

* library linking fix

* Foundry migrations fix

* migration tests fix

* CI bump

* Little cleanup + retrigger CI

* forgot to commit Validators.sol

* Fixed the ABI encoded

* lint

* Fix contract versions

* add Adapter to ignored contracts

* revert of ReentrancyGuard change

* lint fix

* remove adapters from check

* storage layout fix

---------

Co-authored-by: pahor167 <hornak.pavel@gmail.com>

pahor167 committed:
force release of canary abi

pahor167 committed:
Merge branch 'feat/l2-epoch-system' of https://github.com/celo-org/celo-monorepo into feat/l2-epoch-system

pahor167 committed:
force new canary version

pahor167 committed:
npm tag revert

soloseng committed:
Soloseng/update-l2-getepochnumber-logic (#11195)

* disable getEpochSize on L2

* update registry

* update relevant interfaces

* update contracts with L2 `getEpochNumber()` logic

* update tests with L2 `getEpochNumber()` logic

* ++ TODO

* moved constants to constants file

* updated allocated supply function to handle L1 & L2 cases

* made `epochManager.currenEpochNumber()` private, to avoid returning 0 when not initialized.

* PR feedback

* Passing validators test using mockEpochManager for L2 tests

* clean up

* fixed other failing tests

* using mockEpochManager instead of interface.

Fixed failing tests.

* happy linter

* revert npm tag ∆

soloseng committed:
++ TODO and comment

pahor167 committed:
add score manager to abis

pahor167 committed:
Merge branch 'feat/l2-epoch-system' of https://github.com/celo-org/celo-monorepo into feat/l2-epoch-system

martinvol commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,322 @@
+  }
+
+  function finishNextEpochProcess(
+    address[] calldata groups,
+    address[] calldata lessers,
+    address[] calldata greaters
```
An optimization to have less calldata is to use the index in a list of validators rather than passing the addresses. Would that make sense?

martinvol replied:
we agreed to leave this for the end, as it will potentially take more gas

soloseng committed:
Split initEpochManager Function (#11199)

* ++ contract function

* ++ comment

soloseng committed:
startNextEpochProcess unit & integration test (#11191)

* unit test with mocks

* ++ integration tests

* clean up

* -- logging

* removed duplicate interface

* using `MockCeloToken` to get test to pass.

Fails when it hits a precompile in `EpochRewards.sol`

* removed endEpochTimestamp

* moved IEpochManager to 0.5 folder

* added L2 conditions for EpochRewards functions using precompiles

Still missing tests

* renamed EpochManagerInitializer due to name conflict

* ++ more unit test

* setup anvil migration
fix name conflict

* compiles

* ++ require fund in unreleased treasury

* Updated regex

* ++ registry 0.8 for testing only

* clean up

* ++ unit test

* initial integration test using L1 devchain

* ++ comment

* -- forge based integration test

* ++ to const

* happy linter

* update contract name

* ++ PR feedback

* ++ checks

* updated carbon address

* proxy stableToken mint call via Validators contract

* -- duplicate imports

* removed registry08. replaced with vm call

* PR feedback

* -- coment

* passing unit tests

* clean up

* ++ mintStable test

* -- TODO; compiles test when filtering

* PR feedback

* updated migration script to add more validators

* passing integration test

* removed test for zero amount

* yarn build fix

* clean up comments && TODO

* revert change as out of scope

pahor167 committed:
E2E EpochManager test + Epoch truffle migrations & Anvil L2 migration build fix (#11198)

soloseng committed:
Soloseng/dynamically-fetch-epochmanagerenabler-address (#11207)

* dynamically fetch epochManagerEnabler && carbonOffsettingPartner

++ to registry

* PR feedback

* removed onlyL1 modifier in setter functions

* updated unit test to reflect changes

* fixing tests

* fix test

* fixed migration data

* fixed migration script error

* removed unused modifier

* removed duplicate or unused code

mcortesi commented on packages/protocol/contracts-0.8/common/ScoreManager.sol:
```diff
@@ -0,0 +1,50 @@
+import "@openzeppelin/contracts8/access/Ownable.sol";
+
+contract ScoreManager is Initializable, Ownable {
+  mapping(address => uint256) public scores;
```
i would have 'validatorScores' and 'groupScores'

It's just weird to do `setValidatorScore(0x1, someScore)` and then do `getGroupScore(0x01)` and get that score.

martinvol replied:
I would actually abstract this, you set scores of addresses, doesn't really matter if it's a group or a validator.

In fact, we have logic to define the score of a group from the score of its members in `Validators.calculateGroupEpochScore`. This function is no longer used because it receives uptimes as parameters, but the logic that I paste could be reused:

```
for (uint256 i = 0; i < uptimes.length; i = i.add(1)) {
      sum = sum.add(FixidityLib.wrap(calculateEpochScore(uptimes[i])));
    }
    return sum.divide(FixidityLib.newFixed(uptimes.length)).unwrap();
 ```
 
 Then we just check if the score manager overwrote that, or the average of its members should be used.

mcortesi replied:
I'm ok with having a `getScore(address)` and then just a `scores` mapping. Or having `getValidatorScore` and `getGroupScore` and then 2 mappings. Either way

soloseng replied:
addressed in #11223 commit 29f7acd23152c1b0e2ec722d0d783bcf1e5bbe0b

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManagerEnabler.sol:
```diff
@@ -0,0 +1,90 @@
+    return (1, 1, 0, 0);
+  }
+
+  function _getFirstBlockOfEpoch(uint256 currentEpoch) internal view returns (uint256) {
```
many ways to do this, there's a previous comment for this. please change

martinvol replied:
this is a pending known issue I'll work on tomorow

soloseng replied:
addressed by https://github.com/celo-org/celo-monorepo/pull/11210

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManagerEnabler.sol:
```diff
@@ -0,0 +1,90 @@
+
+    for (uint256 i = 0; i < numberElectedValidators; i++) {
+      // TODO: document how much gas this takes for 110 signers
+      address validatorAccountAddress = getAccounts().validatorSignerToAccount(
```
cache `getAccounts()` in a variable, to avoid fetching it each time

soloseng replied:
addressed by https://github.com/celo-org/celo-monorepo/pull/11215

mcortesi commented on packages/protocol/contracts-0.8/common/CeloUnreleasedTreasure.sol:
```diff
@@ -0,0 +1,64 @@
+   */
+  function release(address to, uint256 amount) external onlyEpochManager {
+    require(address(this).balance >= amount, "Insufficient balance.");
+    require(getCeloToken().transfer(to, amount), "CELO transfer failed.");
```
any reason why we prefer doing an ERC20 transfer, instead of a native transfer. My rule of thumb is that Token Duality is only for cases where you NEED a ERC20, for the rest of use cases we should just use native transfer. They are much cheaper gas wide

martinvol replied:
It's nice to have the event log and the checks the transfer function has, as rule of thumb I always use the ERC20 transfer 🤔

mcortesi replied:
yo do have an event here. The `Release` event.

martinvol replied:
I meant in general, I find it nice to always leave a trace with Celo transfers

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+
+  modifier onlyEpochManagerEnabler() {
+    require(
+      msg.sender == registry.getAddressForOrDie(EPOCH_MANAGER_ENABLER_REGISTRY_ID),
```
I'm against having a registry entry for the epoch manager enabler. This is too much complexity for something that it's just a one time only contract. We can just inject the enable address at initialize, and call it a day

soloseng replied:
Not sure why you say its complex. Our current tooling already adds the contract to the registry when its deployed. 

If we were to pass the epochManagerEnabler address during initialization of the epochManager, we would have to modify our deployment tooling, or have to deploy in two phases just to add this address. That seems more complex then just fetching from the registry.

Could you elaborate more on the complexity you see here?

mcortesi replied:
What i mean is that you're adding a registry entry (that lives forever) for a smart contract that it's basically a script that is only applied once during migration.

That sound "smelly". For a script that i'm only using once, i want to forget about it the moment i'm done using it. Instead, in 2 years, you'll see the registry and there it is. And then on all mappings of registry entries you'll have it.

So, you're choosing something that it's easy now, in expense of keeping garbage for the whole history of the celo chain.

martinvol replied:
I agree with seb, it does have more moving pieces but saves having to do follow-up calls to the contract after it was deployed.

mcortesi replied:
primitives team decision. My take still the same, leaving trails and living objects feels smelly to me. Less pain today in exchange for more pain tomorrow

martinvol replied:
Can we resolve this @soloseng ?

soloseng replied:
yes, leaving as is (with call to registry)

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+      address(registry.getAddressForOrDie(CELO_UNRELEASED_TREASURE_REGISTRY_ID)).balance > 0,
+      "CeloUnreleasedTreasury not yet funded."
+    );
+    require(
```
look like this condition should come before the one above

martinvol replied:
aren't

```
    require(
      address(registry.getAddressForOrDie(CELO_UNRELEASED_TREASURE_REGISTRY_ID)).balance > 0,
      "CeloUnreleasedTreasury not yet funded."
    );
 ```

and 

```
     require(
      getCeloToken().balanceOf(registry.getAddressForOrDie(CELO_UNRELEASED_TREASURE_REGISTRY_ID)) >
        0,
      "CeloUnreleasedTreasury not yet funded."
    );
 ```
 the same check?

mcortesi replied:
i guess one was checking there's a registry address, and the other that i has a non zero balance. but i agree can be done in on step

soloseng replied:
updated

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+
+    Epoch storage _currentEpoch = epochs[currentEpochNumber];
+    _currentEpoch.firstBlock = firstEpochBlock;
+    _currentEpoch.startTimestamp = block.timestamp;
```
haven't we said that timestamp is a parameter to this function?

martinvol replied:
why would it be different that the current block?

mcortesi replied:
same reason why the firstblock of epoch is no the current one. The timestamp should be the timestamp of that block (an approximation of it)

soloseng replied:
Calculating the first block timestamp (or approximation) would be more complex than getting the first block of an epoch, given that the block time is not a constant like the epoch size. We could spend time to find a way to make it happen, but I'd suggest doing so for mainnet launch if we want to go that route.

As of now, the decision was as follows:

> // DESIGNDESICION(XXX): we assume that the first epoch on the L2 starts as soon as the system is initialized
  // to minimize amount of "limbo blocks" the network should stop relatively close to an epoch number (but wigh enough time)
  // to have time to call the function EpochInitializer.migrateEpochAndValidators()

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+
+  /// returns the current epoch Info
+  function getCurrentEpoch() external view returns (uint256, uint256, uint256, uint256) {
+    Epoch storage _epoch = epochs[currentEpochNumber];
```
to be consistent, we should add:
```
require(systemAlreadyInitialized(), "EpochManager system not yet initialized.");
```
probably just have a modifier for that

soloseng replied:
addressed by https://github.com/celo-org/celo-monorepo/pull/11223 commit b4ddea75

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+   * @dev Can only be set by owner.
+   */
+  function setEpochDuration(uint256 newEpochDuration) public onlyOwner {
+    epochDuration = newEpochDuration;
```
can i change epoch duration in the middle of epoch change? That is if `isOnEpochProcess() == True`?

martinvol replied:
nit catch, safer not to. and btw this setter is missing an event

soloseng replied:
addressed by #11223 commit 4e4decfe76a55889426a596e0173dd371a250f81

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+  }
+
+  function systemAlreadyInitialized() public view returns (bool) {
+    return initialized && elected.length > 0;
```
since `elected.length` is check on initialziation, and i guess we won't change epoch if the new election give cero elected. we maybe should remove that part of the condition?

soloseng replied:
address by https://github.com/celo-org/celo-monorepo/pull/11223 in commit 9ac350a2c861543af55172f480c9ddd5d6caa250

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+    require(isOnEpochProcess(), "Epoch process is not started");
+    // finalize epoch
+    // last block should be the block before and timestamp from previous block
+    epochs[currentEpochNumber].lastBlock = block.number - 1;
```
haven't we said that we might just remove `lastBlock` as it can be computed from `firstBlock` of next epoch?

soloseng replied:
I think the consensus was to remove the `lastBlockTimestamp` since it's not possible to get historical timestamp.

We could see about removing it if you think it's an issue, but I think that getting the result without having to make extra computation is neat.

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+    }
+    // Mint all cUSD required for payment and the corresponding CELO
+    validators.mintStableToEpochManager(totalRewards);
+    // this should have a setter for the oracle.
```
?

martinvol replied:
this is an early comment that got overlooked, I meant to decouple SortedOracles in this contract. So instead of doing

`IOracle(address(getSortedOracles()))` it should be doing `getcUSDOracle()`

```
function getcUSDOracle() {
   return oracleAddress;
}
```

and `oracleAddress` should be set in the initializer (and then pointed to SortedOracles during the deployment).

martinvol replied:
the address should probably be set in the `initializer` function.
Either via a parameter or fetching sortedOracles

soloseng replied:
addressed in #11223 commit bb38303e5bd9b8aa807aa7531460c550d14b73d8

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+      totalRewards += validatorReward;
+    }
+    // Mint all cUSD required for payment and the corresponding CELO
+    validators.mintStableToEpochManager(totalRewards);
```
why is the validator minting and not us?

To me, as EpochManager is the one minting CELO to compensate for the cUSD minted, then it should be doing both action, not only one

martinvol replied:
the reason is that Validators.sol is already whitelisted on Mento tokens for minting, and it's hardcoded. So to avoid having to upgrade Mento contracts and coordinate with them, better to just proxy the call using validators.

soloseng replied:
addressed by https://github.com/celo-org/celo-monorepo/pull/11223 commit https://github.com/celo-org/celo-monorepo/commit/66c4a89542ac3b03a3dad2fadcac803f7bba3205

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+    epochs[currentEpochNumber].startTimestamp = block.timestamp;
+
+    for (uint i = 0; i < elected.length; i++) {
+      address group = getValidators().getValidatorsGroup(elected[i]);
```
we can cache `getValidators()`, `getScoreReader()` and `getElection()`

martinvol replied:
at first I thought mariano was saying at contract level, then I realized he just means outside of the loop

soloseng replied:
addressed by #11223 commit 1b37ffd1655f720283510eebdab5ea006fb2e772

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+    for (uint i = 0; i < elected.length; i++) {
+      address group = getValidators().getValidatorsGroup(elected[i]);
+      if (!processedGroups[group].processed) {
+        epochProcessing.toProcessGroups++;
```
given that `toProcessGroups` is only ever used on this function and then cleared. We might just use a local variable and save us from storage access

soloseng replied:
addressed in #11223 commit 1ced110faa9a78bf1372d5fc9232dead1b9190da

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+      uint256 i = ii - 1;
+      ProcessedGroup storage processedGroup = processedGroups[groups[i]];
+      // checks that group is actually from elected group
+      require(processedGroup.processed, "group not processed");
```
better to say "group not from current elected set"

soloseng replied:
addressed by #11223 commit cbf5fdc59053c4a5162354dcb6058476e276706e

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+      // by doing this, we avoid processing a group twice
+      delete processedGroups[groups[i]];
+    }
+    getCeloUnreleasedTreasure().release(
```
this and next call could happen on the "start()" epoch process method, and with that we save ourselves from keeping the value on `epochProcessing`

martinvol replied:
doing so would change how much is to be minted, so if done then should be done before  `getEpochRewards().calculateTargetEpochRewards()`, and thus ` getEpochRewards().calculateTargetEpochRewards()` would have to be called twice in start function

mcortesi replied:
not really. `calculateTargetEpochRewards()` is called on `startNextEpochProcess()`.

My suggestion was that just after that call, do the `release()` to the community fund and carbon fund, instead of saving those on storage variables and doing them here.

soloseng replied:
We could move the release() call to the community fund and carbon fund in the `startEpochProcess()`, but that means there would be no straight forward way to fetch what was distributed to the community or carbon fund while the epoch is being processed. 

This generally wouldn't be an issue, but since the epoch processing is split into 2 functions that are likely to be called on separate blocks, I think it would be good to receive all the processing information when calling `getEpochProcessingState()` while an epoch is still being processed.

mcortesi commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,338 @@
+    );
+    // run elections
+    elected = getElection().electValidatorSigners();
+    // TODO check how to nullify stuct
```
TODO here!

mcortesi commented on packages/protocol/contracts-0.8/governance/Validators.sol:
```diff
@@ -613,13 +644,26 @@ contract Validators is
+   * @notice Allows the EpochManager contract to mint stable token for itself.
+   * @param amount The amount to be minted.
+   */
+  function mintStableToEpochManager(
```
as said earlier i think this is unnecessary delegation just have this logic on the epoch manager

soloseng replied:
The original approach was to have the logic in `epochManager` , but for this to work, we would need mento to update their contract to allow `epochManager` to mint. Since they have a different deployment/development process, this would add more time to release.

mcortesi replied:
ok, then please add a comment here, explaining this, so we can change it in the future

soloseng replied:
addressed by https://github.com/celo-org/celo-monorepo/pull/11223 commit 66c4a89542ac3b03a3dad2fadcac803f7bba3205

mcortesi commented on packages/protocol/contracts-0.8/governance/Validators.sol:
```diff
@@ -1325,11 +1420,17 @@ contract Validators is
   function updateMembershipHistory(address account, address group) private returns (bool) {
     MembershipHistory storage history = validators[account].membershipHistory;
-    uint256 epochNumber = getEpochNumber();
+    uint256 epochNumber;
```
call `_getEpochNumber()`

soloseng replied:
addressed by #11223 commit ef98ba53ff8b14c25ede6d8860e4339f798337d0

mcortesi commented on packages/protocol/contracts/common/interfaces/IEpochManager.sol:
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.5.13 <0.9.0;
+
+interface IEpochManager {
```
review methods here, i don't see all of them listed

soloseng replied:
addressed by #11223 commit e94259d57e00e46593812fcbdb619f6baf9333fd

mcortesi commented on packages/protocol/contracts/common/interfaces/IEpochManagerEnabler.sol:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.5.13 <0.9.0;
+
+interface IEpochManagerEnabler {
```
unnecessary to have an interface for this

soloseng replied:
This is used for forge testing and migration scripts.

mcortesi commented on packages/protocol/contracts/common/interfaces/IScoreManager.sol:
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.5.13 <0.9.0;
+
+interface IScoreManager {
```
as said in some other PR. A Score manager has getters for scores and only that. Setters are an implemetation detail, you can create another socre manager that computes scores some other way and setters wouldn't make sense there.

Instead just have some IScoreManager or IScoreREader, and that's the dependency EpochManager needs

martinvol replied:
in this case maybe move to another interface `IScoreSetter` or maybe for now not needed at all.

soloseng replied:
resolving this, as epochManager already uses `IScoreReader`

m-chrzan committed:
Implement sending of allocated validator payments (#11197)

pahor167 committed:
EpochManager fixes (#11208)

* truffle build fix

* build fix

* PR comments

* prettify

* rename of registerValidator overload

* bug fix

* extensing epochManager e2e test

* yarn lint

m-chrzan committed:
Reset pending payment to 0 after sending (#11209)

Reset pending payment to 0 when sending

martinvol commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,415 @@
+   *   delegation beneficiary.
+   * @param validator Account of the validator.
+   */
+  function sendValidatorPayment(address validator) external {
```
`nonReentrant`?

m-chrzan replied:
We could, for an extra extra sanity check, but we're only interacting with other core contracts and not doing any native transfers.

martinvol replied:
Mento is not a core contract anymore

soloseng replied:
addressed by https://github.com/celo-org/celo-monorepo/pull/11217

martinvol commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -0,0 +1,415 @@
+      .fromFixed();
+    uint256 validatorPayment = remainingPayment.fromFixed() - delegatedPayment;
+
+    IStableToken stableToken = IStableToken(getStableToken());
```
I think we should use IER20 here to decouple

martinvol committed:
Deleted duplicated import

martinvol committed:
Make captureEpochAndValidators work in constant time (#11210)

m-chrzan committed:
Use more general interface for token transfer (#11216)

soloseng committed:
Epoch manager enabler tests (#11213)

* ++ basic test

* -- celoToken balance check

* cleanup comments

* use celoToken instead of native token for `initializeSystem` balance check

* ++ more test

* removed additional epochs

m-chrzan committed:
Make sendValidatorPayment nonReentrant (#11217)

martinvol committed:
Martinvol/return account instead of signer (#11215)

pahor167 committed:
EpochManager e2e tests add/remove validators between ecpochs (#11214)

m-chrzan committed:
Send validator payment on actions that would modify reward distribution (#11211)

soloseng committed:
Rename `CeloUnreleasedTreasure` to `CeloUnreleasedTreasury` (#11220)

* updated `CeloUnrealeasedTreasure` to `CeloUnreleasedTreasury`

* update `CeloUnreleasedTreasury` initial balance to use defined constant

soloseng committed:
renamed file

pahor167 committed:
Update score manager behaviour (#11224)

* Update score manager behavior

* scoremanager events

* Remove magic number

---------

Co-authored-by: Martín Volpe <martin.volpe@gmail.com>

soloseng committed:
PR-feedback (#11223)

soloseng committed:
L1-L2-checks (#11225)

martinvol committed:
Merge branch 'release/core-contracts/12' of github.com:celo-org/celo-monorepo into feat/l2-epoch-system
